frame/evm/src/precompiles.rs

```diff
@@ -0,0 +1,78 @@
+use sp_std::vec::Vec;
+use sp_core::H160;
+use evm::{ExitError, ExitSucceed};
+
+/// Custom precompiles to be used by EVM engine.
+pub trait Precompiles {
+	/// Try to execute the code address as precompile. If the code address is not
+	/// a precompile or the precompile is not yet available, return `None`.
+	/// Otherwise, calculate the amount of gas needed with given `input` and
+	/// `target_gas`. Return `Some(Ok(status, output, gas_used))` if the execution
+	/// is successful. Otherwise return `Some(Err(_))`.
+	fn execute(
+		address: H160,
+		input: &[u8],
+		target_gas: Option<usize>,
+	) -> Option<core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>>;
+}
+
+impl Precompiles for () {
+	fn execute(
+		_address: H160,
+		_input: &[u8],
+		_target_gas: Option<usize>
+	) -> Option<core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>> {
+		None
+	}
+}
+
+/// One single precompile used by EVM engine.
+pub trait Precompile {
+	/// Try to execute the precompile. Calculate the amount of gas needed with given `input` and
+	/// `target_gas`. Return `Ok(status, output, gas_used)` if the execution is
+	/// successful. Otherwise return `Err(_)`.
+	fn execute(
+		input: &[u8],
+		target_gas: Option<usize>,
+	) -> core::result::Result<(ExitSucceed, Vec<u8>, usize), ExitError>;
+}
+
+macro_rules! impl_precompile_tuple {
```

Comment:
considered using impl_trait_for_tuples?

Reply:
Yes, please use `impl_trait_for_tuples` for this!

Reply:
Just search the code base for this.

Reply:
Fixed!